tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/internal/c_api/AbstractTF_Status.java

```diff
@@ -69,17 +75,17 @@ public void throwExceptionIfNotOK() {
         case TF_OK:
           break;
         case TF_INVALID_ARGUMENT:
-          throw new IllegalArgumentException(TF_Message(s).getString());
+          throw new TFInvalidArgumentException(TF_Message(s).getString());
         case TF_UNAUTHENTICATED:
         case TF_PERMISSION_DENIED:
-          throw new SecurityException(TF_Message(s).getString());
+          throw new TFPermissionDeniedException(TF_Message(s).getString());
         case TF_RESOURCE_EXHAUSTED:
         case TF_FAILED_PRECONDITION:
-          throw new IllegalStateException(TF_Message(s).getString());
+          throw new TFFailedPreconditionException(TF_Message(s).getString());
         case TF_OUT_OF_RANGE:
-          throw new IndexOutOfBoundsException(TF_Message(s).getString());
+          throw new TFOutOfRangeException(TF_Message(s).getString());
         case TF_UNIMPLEMENTED:
-          throw new UnsupportedOperationException(TF_Message(s).getString());
+          throw new TFUnimplementedException(TF_Message(s).getString());
```

Comment:
Good point, I had created `TFUnauthenticatedException` and `TFResourceExhaustedException` as well but hadn't mapped to them. Just added them to this mapping.

I think it's a good idea to change it for all exceptions. The ambiguity issue we found with `IndexOutOfBoundsException` applies to all of these exceptions. There weren't too many issues in tests to address so within our codebase, this is a pretty low-touch change.

Are there other major cases where current users might be catching errors from `tensorflow-core`? My guess is that not *too* many users will depend strongly on this. If we document this change well during our alpha releases, I think it can be adjusted to.

Reply:
I've gone back and forth on this, and I don't have strong opinions about if the others should be TF exceptions or not. The main reason that I want the `TFOutOfRangeException` is because we've made that part of the API that the user has to know about and catch. As far as I can tell the rest of the errors are programmer errors that are reasonable to leave as the standard exceptions, but they are probably also fine as wrapped TF exceptions so we have more of an idea of the provenance. I agree with Karl that if we're going to map most of them we should map all of them for completeness.

Reply:
Ok guys, sorry but I think I changed a bit my mind on this again... I think it is good to have distinct exception per possible errors thrown by TensorFlow, like you did @dhruvrajan. But I think those exceptions should extend from their matching runtime exception (the one that were thrown before) ratter than from the generic `TensorFlowException`.

My reasoning is that it does not really matter if, for instance, a parameter is detected to be invalid at the JVM level or at the native level for a user. So if we throw a `IllegalArgumentException` when validating user parameters in the JVM, we should throw the same if the validation is done in the native library. 

Still, having concrete type like `TFOutOfRangeException` will allow us to catch explicitly this exception without catching all `IndexOutOfBoundException`, which was the original purpose here.

Reply:
I guess the thing here is, when a user tries to catch an `IndexOutOfBounds` exception to catch they expect from a Java array, for example (in their own code), do we want it to catch other errors thrown from tensorflow-core? Granted, these cases should be pretty infrequent, but it's still a very confusing semantic.